---
title: "Aneurysm_bulk_RNA"
author: "YS"
date: "9/29/2021"
output: html_document
---

```{r 1. setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

#https://2-bitbio.com/post/fuzzy-cmeans-clustering-of-rnaseq-data-using-mfuzz/
#http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#variations-to-the-standard-workflow
```

```{r 2. loading packages}
library(limma)
library("marray")
library(ReactomePA)
library(clusterProfiler)
library(ComplexHeatmap)
library("org.Mm.eg.db")
library(dplyr)
library(RColorBrewer)
library(tibble)
library(ggplot2)
library(Mfuzz)
#library(Glimma)
library(edgeR)
library("vsn")
#library(Mus.musculus)
library(DESeq2)
library("pheatmap")
library("RColorBrewer")
require(statmod)
setwd("~/Documents/work/Aneurysm-Study")
```

```{r 3. loading raw counts}
Feature_count=read.table("Feature_counts_data_matrix.txt",header = T,sep = "\t",stringsAsFactors = F)
Feature_count<-Feature_count[,c(1,2,3,5,4,6,7)]
names(Feature_count)
names(Feature_count)<-c("mgR.1","mgR.2","mgR.C21","mgR.LOS","mgR.C21.LOS","WT.1","WT.2")

meta<-data.frame(run=names(Feature_count),species="mouse")
rownames(meta)<-meta$run
meta$condition<-c("mgR","mgR","mgR.C21","mgR.LOS","mgR.C21.LOS","WT","WT")

lib.size <- estimateSizeFactorsForMatrix(Feature_count)
ed <- t(t(Feature_count)/lib.size)
```

```{r 4. selecting highly variable genes from raw counts} 
###calculate library size then normalize(if not done yet)
# lib.size <- estimateSizeFactorsForMatrix(Feature_count)
# ed <- t(t(Feature_count)/lib.size)
###METHOD 1####
#Calculate estimates of variance, coefficient of variation
means <- rowMeans(ed)
vars <- apply(ed,1,var)
cv2 <- vars/means^2
#Now fit a regression line based on the controls:
minMeanForFit <- unname( quantile( means[ which( cv2 > .3 ) ], .95 ) )
useForFit <- means >= minMeanForFit # & spikeins
fit <- glmgam.fit( cbind( a0 = 1, a1tilde = 1/means[useForFit] ),cv2[useForFit] )
a0 <- unname( fit$coefficients["a0"] )
a1 <- unname( fit$coefficients["a1tilde"])
fit$coefficients
#Now add the fit and the 95% confidence interval to our plot:
xg <- exp(seq( min(log(means[means>0])), max(log(means)), length.out=1000 ))
vfit <- a1/xg + a0
df <- ncol(ed) - 1
#Rank genes by the significance of deviation from the fit
afit <- a1/means+a0
varFitRatio <- vars/(afit*means^2)
varorder <- order(varFitRatio,decreasing=T)
oed <- ed[varorder,]

# save for the next exercise
save(oed,file="oed_standard.RData")
#plotting
pdf("test.pdf")
par(mar=c(3.5,3.5,1,1),mgp=c(2,0.65,0),cex=0.9); smoothScatter(log(means),log(cv2)); lines( log(xg), log(vfit), col="black", lwd=3 ); lines(log(xg),log(vfit * qchisq(0.975,df)/df),lty=2,col="black"); lines(log(xg),log(vfit * qchisq(0.025,df)/df),lty=2,col="black");
# add top 500 genes
points(log(means[varorder[1:500]]),log(cv2[varorder[1:500]]),col=2)
dev.off()
#highly variable genes
# means[varorder[1:500]]

#evaluate statistical significance of the deviation
pval <- pchisq(varFitRatio*df,df=df,lower.tail=F)
adj.pval <- p.adjust(pval,"fdr")
sigVariedGenes <- adj.pval<1e-3;
table(sigVariedGenes)

#Look at how the most variable genes are expressed …
pdf("test.pdf")
m <- oed[1:500,]
heatmap(m/apply(m,1,max),zlim=c(0,1),cluster_rows=T,cluster_cols=T,scale="none")
dev.off()

###METHOD 2   winsorize#####
winsorize <- function (x, fraction=0.05) {
        if(length(fraction) != 1 || fraction < 0 ||
           fraction > 0.5) {
                stop("bad value for 'fraction'")
        }
        lim <- quantile(x, probs=c(fraction, 1-fraction))
        x[ x < lim[1] ] <- lim[1]
        x[ x > lim[2] ] <- lim[2]
        x
}

# winsorize to remove 2 most extreme cells (from each side)
wed <- t(apply(ed, 1, winsorize, fraction=2/ncol(ed)))

# now let's recalculate the most variable genes with the winsorized matrix (wed)
means = rowMeans(wed); vars = apply(wed,1,var); cv2 <- vars/means^2
xg <- exp(seq( min(log(means[means>0])), max(log(means)), length.out=1000 ))
useForFit <- means >= unname( quantile( means[ which( cv2 > .3 ) ], .95 ) ) 
fit <- glmgam.fit( cbind( a0 = 1, a1tilde = 1/means[useForFit] ),cv2[useForFit] )
afit <- fit$coef["a1tilde"]/means+fit$coef["a0"]
vfit <- fit$coef["a1tilde"]/xg+fit$coef["a0"]
varFitRatio <- vars/(afit*means^2)
varorder <- order(varFitRatio,decreasing=T)
oed <- wed[varorder,]
# save for the next exercise
save(oed,file="oed_winsorize.RData")

#plotting
df <- ncol(ed) - 1

pdf("test.pdf")
par(mar=c(3.5,3.5,1,1),mgp=c(2,0.65,0),cex=0.9); smoothScatter(log(means),log(cv2)); lines( log(xg), log(vfit), col="black", lwd=3 ); lines(log(xg),log(vfit * qchisq(0.975,df)/df),lty=2,col="black"); lines(log(xg),log(vfit * qchisq(0.025,df)/df),lty=2,col="black");
# add top 500 genes
points(log(means[varorder[1:500]]),log(cv2[varorder[1:500]]),col=2)
dev.off()

#heatmap of top genes
pdf("test.pdf")
m <- oed[1:500,]
heatmap(m/apply(m,1,max),zlim=c(0,1),cluster_rows=T,cluster_cols=T,scale="none",show_rownames=FALSE)
dev.off()


###METHOD 3 PCA getting top genes####
#(not good)
cols=c("mgR.1"="red","mgR.2"="red","WT.1"="blue","WT.2"="blue","mgR.C21"="yellow","mgR.C21.LOS"="green","mgR.LOS"="purple")
require(pcaMethods)
pcs <- pca(oed,nPcs=5)
plot(loadings(pcs)[,1],loadings(pcs)[,2],xlab="PC1",ylab="PC2",col=cols)
#We can get similarly good separation with just the top 100 most variable genes
pcs <- pca(oed[1:2500,],nPcs=5)
plot(loadings(pcs)[,1],loadings(pcs)[,2],xlab="PC1",ylab="PC2",col=cols)

###comparison between the highly variable genes between 2 method####
load("~/Documents/work/Aneurysm-Study/oed_winsorize.RData")
oed_winsorize<-oed
load("~/Documents/work/Aneurysm-Study/oed_standard.RData")
oed_standard<-oed

gene_w<-rownames(oed_winsorize[1:3500,])
gene_s<-rownames(oed_standard[1:3500,])

length(intersect(gene_w,gene_s))
#341 out of 500+500 (0.682)
#1941 out of 2500+2500 (0.7764)
#2712 out of 3500+3500 3500(0.77)
#http://pklab.med.harvard.edu/scw2014/subpop_tutorial.html
```

```{r 5. selecting highly variabe genes from Deseq2}
dds <- DESeqDataSetFromMatrix(countData = Feature_count,
                              colData = meta,
                              design = ~ condition)
dds$condition <- relevel(dds$condition, ref = "mgR")


#Pre-filtering
keep <- rowSums(counts(dds)) >= 10
dds <- dds[keep,]
dds <- DESeq(dds)
#Generate the  normalized counts
colSums(counts(dds))
dds <- estimateSizeFactors(dds)
sizeFactors(dds)
normalized_counts <- counts(dds, normalized=TRUE)
colSums(counts(dds, normalized=T))

#
vsd <- vst(dds, blind=FALSE)
rld <- rlog(dds, blind=FALSE)
ntd <- normTransform(dds)
# meanSdPlot(assay(ntd))
# meanSdPlot(assay(vsd))
# meanSdPlot(assay(rld))

#sample-to-sample distances(vsd or rld)####
#vsd method####
sampleDists <- dist(t(assay(vsd)))
sampleDistMatrix <- as.matrix(sampleDists)
colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255)
pheatmap(sampleDistMatrix,
         clustering_distance_rows=sampleDists,
         clustering_distance_cols=sampleDists,
         col=colors)
plotPCA(vsd, intgroup=c("condition"))

#rld method####
rld_mat <- assay(rld)
pca <- prcomp(t(rld_mat))
# Create data frame with metadata and PC3 and PC4 values for input to ggplot
df <- cbind(meta, pca$x)
ggplot(df) + geom_point(aes(x=PC3, y=PC4, color = condition))
#Hierarchical Clustering
### Extract the rlog matrix from the object
rld_mat <- assay(rld)    ## assay() is function from the "SummarizedExperiment" package that was loaded when you loaded DESeq2
### Compute pairwise correlation values
rld_cor <- cor(rld_mat)    ## cor() is a base R function
head(rld_cor)   ## check the output of cor(), make note of the rownames and colnames
### Plot heatmap
pheatmap(rld_cor)

#heatmap of selected genes
# These are the genes most variable across all samples regardless of which samples they are.
topVarGenes <- order(rowVars(assay(rld)),decreasing=TRUE)
tmp<-rld@assays@data@listData[[1]]
tmp<-tmp[topVarGenes[1:5000],]

pdf("test.pdf")
pheatmap(tmp, 
         cluster_rows=TRUE, show_rownames=F,
         cluster_cols=TRUE,scale = "row")
dev.off()
```

```{r 6. Mfuzz showing different treatment group trend}
#Replicated arrays in the exprs matrix are treated as independent by the mfuzz function 
#i.e. they should be averaged prior to clustering or placed into different distinct ExpressionSet objects.

##step1: average biological replicate

#normalized_counts from deseq2 result: normalized_counts <- counts(dds, normalized=TRUE)
#ave.A <- rowMeans( counts(dds, normalized=TRUE)[ , dds$condition == "A" ] )
normalized_counts<-as.data.frame(normalized_counts)
normalized_counts <- normalized_counts %>%
        add_column(mgR = (normalized_counts$mgR.1+normalized_counts$mgR.2)/2,
                   .after = "mgR.2") 
normalized_counts <- normalized_counts %>%
        add_column(WT = (normalized_counts$WT.1+normalized_counts$WT.2)/2,
                   .after = "WT.2") 
pheatmap(cor(normalized_counts))#check if the averaged group highly correlated with both replicates
normalized_counts<-normalized_counts[,-c(1,2,7,8)]

##step2: mfuzz pre-processing
normalized_counts <- data.matrix(normalized_counts)
eset <- new("ExpressionSet",exprs = normalized_counts)
# remove low sd genes
eset <- filter.std(eset,min.std=0)
#standardise，
eset <- standardise(eset)
#3.clustering, Mfuzz need 2 arguments，c(the cluster we are aiming for,arbitury)，fuzzifier(m: evaluated by mestimate)
m <- mestimate(eset)
Dmin(eset, m=m, crange=seq(2,22,1), repeats=3, visu=TRUE) #slow help you find out the max cluster number

c <- 10# thins number should be adjusted after reviewing plots

set.seed(1030)
cl <- mfuzz(eset, c = c, m = m)

#4 visualization
#color.2 <- colorRampPalette(rev(c("#ff0000", "Yellow", "OliveDrab1")))(1000)
pdf("mfuzz_s5_clustreing_c10.pdf")
mfuzz.plot(eset,cl,mfrow=c(3,4),new.window= FALSE,time.labels=colnames(normalized_counts))
dev.off()

##alternative visualization
mfuzz.plot2(eset,cl=cl,mfrow=c(3,4),colo="fancy",time.labels=colnames(normalized_counts),
ax.col="red",bg = "white",col.axis="red",col.lab="black",
col.main="black",col.sub="blue",col="blue",cex.main=0.5,cex.lab=1.1)


##5. validation
#One thing worth including in any cluster analysis is a post hoc test of the correlation between cluster centroids.
#If you over clustered, and ended up with redundant clusters, you can detect this by correlating the centers to see if they are too similar.
#Ideally, no two clusters should exhibit a correlation greater than 0.85
cor(t(cl[[1]])) ##pheatmap(cor(t(cl[[1]])))
#extracts membership values 
acore <- acore(eset,cl,min.acore=0)
acore_list <- do.call(rbind, lapply(seq_along(acore), function(i){ data.frame(CLUSTER=i, acore[[i]])}))

#6 expore cl object
# gene number in each cluster
cl$size
# getting the genes in each cluster
c1_gene<-names(cl$cluster[cl$cluster == 1])
# checking the membership of gene and each cluster
cl$membership


### Single cluster  with colorbar (cluster # 3) 
X11(width=12)
mat <- matrix(1:2,ncol=2,nrow=1,byrow=TRUE)
l   <- layout(mat,width=c(5,1))
 
mfuzz.plot2(eset,cl=cl,mfrow=NA,colo="fancy", ax.col="red",bg = "black",col.axis="red",col.lab="white",
 col.main="green",col.sub="blue",col="blue",cex.main=2, single=1,x11=FALSE)

mfuzzColorBar(main="Membership",cex.main=1)
```

```{r 7. Mfuzz cluster1 genes heatmap(DEseq rld slot)}
#heatmap of selected genes
tmp<-rld@assays@data@listData[[1]]
tmp<-tmp[c1_gene,]
pdf("test.pdf")
out <-pheatmap(tmp, 
         cluster_rows=TRUE, show_rownames=F,
         cluster_cols=FALSE,scale = "row")
out
dev.off()

#Re-order original data (genes) to match ordering in heatmap (top-to-bottom)
rownames(tmp[out$tree_row[["order"]],])
#If you want something like gene-to-cluster assignment, you can 'cut' your row dendrogram into a pre-selected number of groups as follows:
#2 groups
sort(cutree(out$tree_row, k=2))
#You can also cut the tree at a pre-defined tree height, and extract the gene-to-cluster assignments at that height:
plot(out$tree_row)
abline(h=4.7, col="red", lty=2, lwd=2)
sort(cutree(out$tree_row, h=4.7))

C1_gene_df<-as.data.frame(sort(cutree(out$tree_row, h=4.7)))
table(C1_gene_df)
C1_gene_conserve<-rownames(subset(C1_gene_df,sort(cutree(out$tree_row, h = 4.7))==1))

pheatmap(tmp[C1_gene_conserve,], 
         cluster_rows=TRUE, show_rownames=F,
         cluster_cols=FALSE,scale = "row")

```

```{r 8. functional annotation}
df <- bitr(C1_gene_conserve,
           fromType = "ENSEMBL",
           toType = c("SYMBOL","ENTREZID"),
           OrgDb = org.Mm.eg.db)

#
kk <- enrichKEGG(gene = df$ENTREZID,organism = 'mmu',
                 pvalueCutoff = 0.7,qvalueCutoff=0.7)
dotplot(kk, showCategory=20,font=7) +
        ggtitle("KEGG pathway over-representation analysis for cluster1 genes")+
        theme(plot.title = element_text(size = 10, face = "bold"))

#
ReaP <- enrichPathway(gene = df$ENTREZID, organism="mouse",
                      pvalueCutoff = 0.7,qvalueCutoff=0.7, readable=TRUE)
dotplot(ReaP, showCategory=20,font=7) +
         ggtitle("Pathway enrichment analysis for cluster1")+
        theme(plot.title = element_text(size = 10, face = "bold"))

#
go_enrich_results <- lapply( c('BP','MF','CC') , function(ont) {
  cat(paste('Now process ',ont ))
  ego <- enrichGO(gene          = df$ENTREZID,
                  OrgDb         = org.Mm.eg.db,
                  ont           = ont ,
                  pAdjustMethod = "BH",
                  pvalueCutoff  = 0.9,
                  qvalueCutoff  = 0.9,
                  readable      = TRUE)

  print( head(ego) )
  dotplot(ego,title=paste0('dotplot_',ont)) %>% print()
  return(ego)
})

write.csv(df,file="cluster1_gene_conserved.csv")

```

```{r orgnizing cluster specific gene expression file}
#
mfuzz_genelist<-list()
out_heatmap<-list()
for(i in c(1,4,6)){
        print(i) 
        id<-paste("Cluster",i)      
        mfuzz_genelist[[id]]<-names(cl$cluster[cl$cluster == i])    
        tmp<-rld@assays@data@listData[[1]]
        tmp<-tmp[mfuzz_genelist[[id]],]
        pdf(paste("heatmap_all_cluster",i,".pdf"))
        out_heatmap[[id]] <-pheatmap(tmp, 
                                     cluster_rows=TRUE, show_rownames=F,
                                     cluster_cols=FALSE,scale = "row")
        out_heatmap[[id]]
        dev.off() 
        
}
mfuzz_genelist$`Cluster 1`<-C1_gene_conserve
#
mfuzz_genelist_annodf<-list()
for(i in names(mfuzz_genelist)){
        print(i)
           
           mfuzz_genelist_annodf[[i]]<- bitr(mfuzz_genelist[[i]],
           fromType = "ENSEMBL",
           toType = c("SYMBOL","ENTREZID"),
           OrgDb = org.Mm.eg.db)
           
           print(length(mfuzz_genelist_annodf[[i]]$ENSEMBL))
           print(length(unique(mfuzz_genelist_annodf[[i]]$ENSEMBL)))
          #rownames(mfuzz_genelist_df[[i]]) <-mfuzz_genelist_df[[i]]$ENSEMBL
}
#
mfuzz_genelist_norm_expr<-list()
for(i in names(mfuzz_genelist)){
        print(i)        
        normalized_counts <- counts(dds, normalized=TRUE)
        mfuzz_genelist_norm_expr[[i]]<- normalized_counts [mfuzz_genelist[[i]],]     
        mfuzz_genelist_norm_expr[[i]]<-as.data.frame(mfuzz_genelist_norm_expr[[i]])
        mfuzz_genelist_norm_expr[[i]]$ENSEMBL<-rownames(mfuzz_genelist_norm_expr[[i]])
        mfuzz_genelist_norm_expr[[i]] <- merge(mfuzz_genelist_norm_expr[[i]],mfuzz_genelist_annodf[[i]],by="ENSEMBL")
        mfuzz_genelist_norm_expr[[i]]<-data.frame(lapply(mfuzz_genelist_norm_expr[[i]], function(y) if(is.numeric(y)) round(y, 2) else y))
        file.name=paste("Cluster",i,"Normalized gene count.csv")
        write.csv(mfuzz_genelist_norm_expr[[i]],file=file.name)
}
  pheatmap(mfuzz_genelist_norm_exp$`Cluster 6` [,2:8],
  cluster_rows=TRUE, show_rownames=F,
  cluster_cols=FALSE,scale = "row")






```

